### PR TITLE
Enable register native with iwasm

### PR DESCRIPTION
**Pull request by wenyongh**
Enable register native with iwasm application, add sample and update document.

### CODE REVIEW

File: samples/native-lib/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    *p_module_name = "env";
+    *p_native_symbols = native_symbols;
+    return sizeof(native_symbols) / sizeof(NativeSymbol);
+}
+```
+
+## Preparation
+
+Please install WASI SDK, download the [wasi-sdk release](https://github.com/CraneStation/wasi-sdk/releases) and extract the archive to default path `/opt/wasi-sdk`.
+And install wabt, download the [wabt release](https://github.com/WebAssembly/wabt/releases) and extract the archive to default path `/opt/wabt`
```
Comment (xujuntwt95329) [inline samples/native-lib/README.md:34]: Seems `wabt` is not required by this sample?

File: samples/native-lib/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# "native-lib" sample introduction
+
+This sample demonstrates how to write required interfaces in native library, build it into a shared library and
+register the shared library to iwasm.
```
Comment (xujuntwt95329) [inline samples/native-lib/README.md:4]: unnecessary line break

**Review by xujuntwt95329 — COMMENTED** [review]
LGTM

File: samples/native-lib/test_add.c [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{
+    return x + y;
+}
+
+/* clang-format off */
+#define REG_NATIVE_FUNC(func_name, signature) \
+    { #func_name, func_name##_wrapper, signature, NULL }
+
+static NativeSymbol native_symbols[] = {
+    REG_NATIVE_FUNC(test_add, "(ii)i")
```
Comment (lum1n0us) [inline samples/native-lib/test_add.c:22]: there should be a function named `test_add`, right?

**Comment by lum1n0us** [comment]
I got a rough idea about "register native". 

Some observations:

- the module name and the signature information of a register function depend on users' input
- users' compilers make sure wrappers are in the library

So, if I can replace the `REG_NATIVE_FUNC` with a .json file, and provide information including the module name, wrappers' names and signatures, wrappers' target,  and plus modifications on `load_and_register_native_libs`, we can let users register existed native function without recompiling their libraries.

Further, if users' can make sure all a WASM module's native dependencies are in existed libraries, and all import entries are using libraries name as modules name and using wasm function names as native function name, we can scan wasm import entries, and create such a .json automatically,

File: samples/native-lib/test_add.c (hunk 3 above)
Comment (wenyongh) [reply samples/native-lib/test_add.c:22]: Yes, see file samples/native-lib/wasm-app/main.c: https://github.com/bytecodealliance/wasm-micro-runtime/pull/1120/files#diff-bc10ed7477a7549d93f8aca9e41f1fb832c68d4939f955bfe9f29c9769a521ff
it is compiled into wasm file, which calls to the imported function test_add/test_sqrt in its main function.

File: samples/native-lib/README.md (hunk 1 above)
Comment (wenyongh) [reply samples/native-lib/README.md]: Done

File: samples/native-lib/README.md (hunk 2 above)
Comment (wenyongh) [reply samples/native-lib/README.md]: Done

File: product-mini/platforms/posix/main.c [hunk 4]
```diff
@@ -8,6 +8,9 @@
 #endif
 #include <stdlib.h>
 #include <string.h>
+#if defined(_POSIX_SOURCE) || defined(_POSIX_C_SOURCE)
```
Comment (yamt) [inline product-mini/platforms/posix/main.c:11]: what's the intention of this `defined(_POSIX_SOURCE) || defined(_POSIX_C_SOURCE)`?
is it meant to be an availability check of dlopen?

on nuttx, the availability of dlopen depends on a build-time config.
i feel it's more straightforward to control this with BH_HAS_DLFCN or something like that.
how do you think?

Comment (yamt) [reply product-mini/platforms/posix/main.c:11]: https://github.com/bytecodealliance/wasm-micro-runtime/pull/1134

**Pull request merged**